docker/daemon.go

```diff
@@ -119,40 +151,52 @@ func mainDaemon() {
 		serveAPIWait <- nil
 	}()
 
-	registryService := registry.NewService(registryCfg)
-	d, err := daemon.NewDaemon(daemonCfg, eng, registryService)
-	if err != nil {
-		eng.Shutdown()
-		logrus.Fatalf("Error starting daemon: %v", err)
-	}
-
-	if err := d.Install(eng); err != nil {
-		eng.Shutdown()
-		logrus.Fatalf("Error starting daemon: %v", err)
-	}
-
-	logrus.Info("Daemon has completed initialization")
-
-	logrus.WithFields(logrus.Fields{
-		"version":     dockerversion.VERSION,
-		"commit":      dockerversion.GITCOMMIT,
-		"execdriver":  d.ExecutionDriver().Name(),
-		"graphdriver": d.GraphDriver().String(),
-	}).Info("Docker daemon")
+	signal.Trap(func() {
+		api.Close()
+		<-serveAPIWait
+		shutdownDaemon(d, 15)
+		if pfile != nil {
+			if err := pfile.Remove(); err != nil {
+				logrus.Error(err)
+			}
+		}
+	})
 
 	// after the daemon is done setting up we can tell the api to start
 	// accepting connections with specified daemon
 	api.AcceptConnections(d)
 
 	// Daemon is fully initialized and handling API traffic
-	// Wait for serve API job to complete
+	// Wait for serve API to complete
 	errAPI := <-serveAPIWait
-	eng.Shutdown()
+	shutdownDaemon(d, 15)
 	if errAPI != nil {
+		if pfile != nil {
+			if err := pfile.Remove(); err != nil {
+				logrus.Error(err)
+			}
+		}
 		logrus.Fatalf("Shutting down due to ServeAPI error: %v", errAPI)
 	}
 }
 
+// shutdownDaemon just wraps daemon.Shutdown() to handle a timeout in case
+// d.Shutdown() is waiting too long to kill container or worst it's
+// blocked there
+func shutdownDaemon(d *daemon.Daemon, seconds time.Duration) {
+	ch := make(chan struct{})
+	go func() {
+		d.Shutdown()
+		close(ch)
+	}()
+	select {
+	case <-ch:
+		logrus.Debug("Clean shutdown succeded")
+	case <-time.After(seconds * time.Second):
+		logrus.Error("Force shutdown daemon")
```

Comment:
@cpuguy83 Who should we kill? :) if `mainDaemon` function exits, then process just stops, nothing can stand this.